### PR TITLE
refactor(proto)!: update `SequencerBlockHeader` and related proto types to not use cometbft header

### PR DESCRIPTION
**Pull request by noot**
## Summary
update `SequencerBlockHeader`, `SequencerBlock`, `FilteredSequencerBlock`, and `CelestiaSequencerBlob` to not contain an entire cometbft header, but only to contain relevant fields required. 

## Background
this is required for the update to ABCI v0.38, as `finalize_block` no longer passes in the entire block header (like `begin_block` did). we keep needed fields like height, chain_id, timestamp, proposer_address, data_hash in the `SequencerBlockHeader`.

## Changes
- update `SequencerBlockHeader` to not contain an entire cometbft header but have height, chain_id, timestamp, proposer_address, data_hash - all fields are now explicitly needed by other components of the stack
- update `SequencerBlock`, `FilteredSequencerBlock`, and `CelestiaSequencerBlob` to have `block_hash` as this can no longer be calculated from the cometbft header as it doesn't exist anymore
- deprecate `rollup_ids_root` in `SequencerBlockHeader` as it can be calculated by the list of rollup IDs in a block

## Testing
unit tests

## Breaking Changelist
- the `SequencerBlockHeader`, `SequencerBlock`, `FilteredSequencerBlock`, and `CelestiaSequencerBlob` are changed.

## Related Issues

 related to  #679

### CODE REVIEW

File: crates/astria-core/src/sequencer/v1/block/mod.rs [hunk 1]
```diff
@@ -327,25 +337,41 @@ enum SequencerBlockErrorKind {
...
 }
 
 #[derive(Debug, Clone, PartialEq)]
 pub struct SequencerBlockHeader {
     // the cometbft header for this sequencer block
-    cometbft_header: tendermint::block::header::Header,
+    // cometbft_header: tendermint::block::header::Header,
+    chain_id: tendermint::chain::Id,
+    height: tendermint::block::Height,
+    time: Time,
```
Comment (SuperFluffy) [inline crates/astria-core/src/sequencer/v1/block/mod.rs:354]: If we redo these, can't we get rid of tendermint here for good? I am looking at the changed `SequencerBlockHeader` in the redefined protos, and these don't show up at the tendermint level anymore? Would much rather use native protos (or well known types).

**Review by SuperFluffy — COMMENTED** [review]
I think making changes on top of the `v1` sequencer API might make maintaining this a bit of a nightmare if we need extra changes on `main`.

Given that we consider changes to the sequencer proto specs breaking (as in, breaking at the network level even though spec evolution is not breaking for protobus), I think this warrants calling this a `v2alpha1`. It will also make it much easier to merge this back into main at a later point.

File: crates/astria-sequencer-relayer/postsubmit.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"state":"fresh"}
```
[the remark this answers is not in the document]
Comment (noot) [reply crates/astria-sequencer-relayer/postsubmit.json:1]: woops

File: crates/astria-conductor/src/conductor.rs [hunk 3]
```diff
@@ -217,7 +217,9 @@ async fn get_sequencer_namespace(client: HttpClient) -> eyre::Result<Namespace>
         .await
         .wrap_err("failed to get block from sequencer after 10 attempts")?;
 
-    Ok(celestia_client::celestia_namespace_v0_from_cometbft_header(
-        block.header().cometbft_header(),
-    ))
+    Ok(
+        celestia_client::celestia_namespace_v0_from_cometbft_chain_id(
+            block.header().chain_id().as_str(),
```
[the remark this answers is not in the document]
Comment (noot) [reply crates/astria-conductor/src/conductor.rs:222]: it only needs the chain id field though, so that seems worse for code understandability to me

File: crates/astria-core/src/sequencer/v1/block/mod.rs (hunk 1 above)
Comment (noot) [reply crates/astria-core/src/sequencer/v1/block/mod.rs:354]: yeah we can - i left them as tendermint as they're originating from a tendermint chain, but can change to our own types also

**Comment by noot** [comment]
@SuperFluffy moved changed protos (block, celestia, and service) to a v2 - ready for rereview!

File: crates/astria-core/src/sequencer/mod.rs [hunk 4]
```diff
@@ -1 +1,39 @@
 pub mod v1;
+
+/// A trait to convert from raw decoded protobuf types to idiomatic astria types.
+///
+/// The primary use of this trait is to convert to/from foreign types.
+pub trait Protobuf: Sized {
```
Comment (noot) [inline crates/astria-core/src/sequencer/mod.rs:6]: I had to move the Protobuf trait into the crates that were implementing it - the problem is that it was implemented on the external `merkle::Proof` type, however we then had two implementations of it (one in `sequencer` and one in `sequencerblock`) which conflicted as the `Raw` types weren't the same.

File: crates/astria-conductor/src/lib.rs [hunk 5]
```diff
@@ -15,7 +15,7 @@ pub mod conductor;
 pub mod config;
 pub(crate) mod executor;
 pub(crate) mod sequencer;
-pub(crate) mod utils;
+mod utils;
```
Comment (joroshiba) [inline crates/astria-conductor/src/lib.rs:18]: why did this publicity here change?

File: crates/astria-core/src/generated/astria.sequencer.v2.rs
Comment (joroshiba) [inline crates/astria-core/src/generated/astria.sequencer.v2.rs:1]: I think the v2 generated bits here should be removed?

File: crates/astria-core/src/lib.rs [hunk 6]
```diff
@@ -9,44 +9,83 @@ pub mod generated;
...
-
-    /// Convert to the raw protobuf type, dropping `self`.
-    ///
-    /// This method provides a default implementation in terms of
-    /// [`Self::to_raw`].
-    fn into_raw(self) -> Self::Raw {
-        Self::to_raw(&self)
-    }
-}
+// /// A trait to convert from raw decoded protobuf types to idiomatic astria types.
```
Comment (joroshiba) [inline crates/astria-core/src/lib.rs:17]: remove?

File: crates/astria-core/src/sequencer/v1/block/mod.rs [hunk 7]
```diff
@@ -14,8 +14,8 @@ use super::{
     IncorrectRollupIdLength,
     RollupId,
 };
-use crate::{
-    sequencer::v1::{
+use crate::sequencer::{
+    v1::{
```
Comment (joroshiba) [inline crates/astria-core/src/sequencer/v1/block/mod.rs:18]: nit: unnecessary change, kinda surprised format doesn't dislike it

File: crates/astria-core/src/sequencer/v1/transaction/action.rs [hunk 8]
```diff
@@ -9,8 +9,8 @@ use penumbra_ibc::IbcRelay;
 use penumbra_proto::penumbra::core::component::ibc::v1::FungibleTokenPacketData;
 
 use super::raw;
-use crate::{
-    sequencer::v1::{
+use crate::sequencer::{
+    v1::{
```
Comment (joroshiba) [inline crates/astria-core/src/sequencer/v1/transaction/action.rs:13]: nit: here as well

File: crates/astria-conductor/src/lib.rs (hunk 5 above)
Comment (noot) [reply crates/astria-conductor/src/lib.rs:18]: just didn't need `pub(crate)`

File: crates/astria-core/src/sequencer/v1/block/mod.rs (hunk 7 above)
Comment (noot) [reply crates/astria-core/src/sequencer/v1/block/mod.rs:18]: what do you mean? this has to be changed because of the protobuf trait change

File: crates/astria-core/src/sequencerblock/v1alpha1/block.rs [hunk 9]
```diff
@@ -0,0 +1,1529 @@
...
+    pub chain_id: tendermint::chain::Id,
+    pub height: tendermint::block::Height,
+    pub time: Time,
+    pub rollup_transactions_root: [u8; 32],
+    pub data_hash: [u8; 32],
+    pub proposer_address: account::Id,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct SequencerBlockHeader {
```
Comment (noot) [inline crates/astria-core/src/sequencerblock/v1alpha1/block.rs:349]: this type is updated

File: crates/astria-core/src/sequencerblock/v1alpha1/block.rs [hunk 10]
```diff
@@ -0,0 +1,1529 @@
...
+    pub rollup_transactions: IndexMap<RollupId, RollupTransactions>,
+    pub rollup_transactions_proof: merkle::Proof,
+    pub rollup_ids_proof: merkle::Proof,
+}
+
+/// `SequencerBlock` is constructed from a tendermint/cometbft block by
+/// converting its opaque `data` bytes into sequencer specific types.
+#[derive(Clone, Debug, PartialEq)]
+#[allow(clippy::module_name_repetitions)]
+pub struct SequencerBlock {
```
Comment (noot) [inline crates/astria-core/src/sequencerblock/v1alpha1/block.rs:552]: this type is updated

File: crates/astria-core/src/sequencerblock/v1alpha1/block.rs [hunk 11]
```diff
@@ -0,0 +1,1529 @@
...
+    pub rollup_transactions_proof: merkle::Proof,
+    // all rollup ids in the sequencer block
+    pub all_rollup_ids: Vec<RollupId>,
+    // proof that `rollup_ids` is included in `data_hash`
+    pub rollup_ids_proof: merkle::Proof,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+#[allow(clippy::module_name_repetitions)]
+pub struct FilteredSequencerBlock {
```
Comment (noot) [inline crates/astria-core/src/sequencerblock/v1alpha1/block.rs:1011]: this type is updated

File: crates/astria-core/src/sequencerblock/v1alpha1/celestia.rs [hunk 12]
```diff
@@ -0,0 +1,616 @@
...
+            rollup_ids,
+            rollup_transactions_proof,
+            rollup_ids_proof,
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+#[allow(clippy::module_name_repetitions)]
+pub struct CelestiaSequencerBlob {
```
Comment (noot) [inline crates/astria-core/src/sequencerblock/v1alpha1/celestia.rs:481]: this type is updated

File: crates/astria-core/src/sequencerblock/v1alpha1/mod.rs [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+use crate::{
+    generated::sequencerblock::v1alpha1 as raw,
+    sequencer::v1::{
+        derive_merkle_tree_from_rollup_txs,
+        RollupId,
+    },
+    sequencerblock::Protobuf,
+};
+
+pub(crate) fn are_rollup_ids_included<'a, TRollupIds: 'a>(
```
Comment (noot) [inline crates/astria-core/src/sequencerblock/v1alpha1/mod.rs:27]: this file is copy pasted from v1

File: crates/astria-celestia-client/src/lib.rs [hunk 14]
```diff
@@ -55,10 +55,10 @@ pub const fn celestia_namespace_v0_from_rollup_id(
 }
 
 #[must_use = "a celestia namespace must be used in order to be useful"]
-pub fn celestia_namespace_v0_from_cometbft_header(header: &tendermint::block::Header) -> Namespace {
+pub fn celestia_namespace_v0_from_cometbft_chain_id(chain_id: &str) -> Namespace {
```
[the remark this answers is not in the document]
Comment (SuperFluffy) [reply crates/astria-celestia-client/src/lib.rs:58]: Raising this again: please actually provide an `chain::Id` argument or rename this ti `celestia_namespace_v0_from_str`. Although I think I would prefer just calling this `celestia_namespace_v0_from_bytes` (as done in #946 )

File: crates/astria-core/src/sequencerblock/mod.rs [hunk 15]
```diff
@@ -0,0 +1,39 @@
+pub mod v1alpha1;
+
+/// A trait to convert from raw decoded protobuf types to idiomatic astria types.
+///
+/// The primary use of this trait is to convert to/from foreign types.
+pub trait Protobuf: Sized {
```
Comment (SuperFluffy) [inline crates/astria-core/src/sequencerblock/mod.rs:6]: Please move this to the top level of `astria-core`

File: crates/astria-core/src/sequencer/mod.rs (hunk 4 above)
Comment (SuperFluffy) [reply crates/astria-core/src/sequencer/mod.rs:6]: Ok, that's not good. We have to find a different solution because that's pretty confusing then. I really don't want to duplicate traits like this.

File: crates/astria-conductor/src/conductor.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (SuperFluffy) [reply crates/astria-conductor/src/conductor.rs]: It's actually a pretty common (and preferred) pattern - law of demeter. You need not know that `block` has a `header` that in turn provides a `chain_id`. It's enough that `block` gives you a chain ID - where it comes from is an implementation detail.

File: crates/astria-core/src/sequencerblock/v1alpha1/mod.rs (hunk 13 above)
Comment (SuperFluffy) [reply crates/astria-core/src/sequencerblock/v1alpha1/mod.rs:27]: Please move this to `lib.rs` or into a utilty module (same for the others) so that we avoid unnecessary code duplication.

File: crates/astria-core/src/sequencerblock/v1alpha1/mod.rs [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+        .audit()
+        .with_root(root)
+        .with_leaf_builder()
+        .write(id.as_ref())
+        .write(&merkle::Tree::from_leaves(rollup_transactions.transactions()).root())
+        .finish_leaf()
+        .perform()
+}
+
+impl Protobuf for merkle::Proof {
```
Comment (SuperFluffy) [inline crates/astria-core/src/sequencerblock/v1alpha1/mod.rs:69]: This is really not ideal. We should either modify the `Protobuf` trait (which I don't like), define newtype wrappers, or have a common `raw::Proof` type.

**Review by SuperFluffy — CHANGES_REQUESTED** [review]
This introduces a lot of (IMO unnecessary) duplicated code. Especially the fact that we have duplicate definintions for `merkle::Proof` and duplicate definitions of the `Protobuf` trait points to the API design not being sound.

**Review by SuperFluffy — APPROVED** [review]
Approving to unblock the move to abci 0.38 and with a nod to #958 that will address my concerns
